File: ese3090/beatpath/beatpath.py
import numpy as np
import itertools


class Profile:
    def __init__(self, profile: np.ndarray):
        self.profile = profile
        self.n_voters = profile.shape[0]
        self.n_candidates = profile.shape[1]

    def get_d_matrix(self):
        """
        Given a profile, return the d matrix.

        Profile is a numpy array of shape (n_voters, n_candidates), where
        each ballot has a ranking of the candidates.

        For example, a ballot [1, 0, 2] means that candidate 0 is in index 1,
        candidate 1 is in index 0, and candidate 2 is in index 2.
        """
        d_matrix = np.zeros((self.n_candidates, self.n_candidates))

        for comb in itertools.combinations(range(self.n_candidates), 2):
            for i in range(self.n_voters):
                a = comb[0]
                b = comb[1]
                if self.profile[i][a] < self.profile[i][b]:
                    d_matrix[a][b] += 1
                else:
                    d_matrix[b][a] += 1

        for i in range(self.n_candidates):
            for j in range(self.n_candidates):
                if i == j:
                    d_matrix[i][j] = 0
                elif d_matrix[i][j] > d_matrix[j][i]:
                    d_matrix[j][i] = 0

        return d_matrix

    def get_p_matrix(self, d_matrix: np.ndarray):
        """
        Given a d matrix, return the p matrix.

        The p matrix is the strength of the strongest path from i to j.

        For example, if the strongest path from i to j is 3, then p_matrix[i][j] = 3.
        """

        p_matrix = np.zeros((self.n_candidates, self.n_candidates))

        for cand1 in range(self.n_candidates):
            for cand2 in range(self.n_candidates):
                if cand1 != cand2:
                    strength = d_matrix[cand1][cand2]
                    if strength > d_matrix[cand2][cand1]:
                        p_matrix[cand1][cand2] = strength

        for cand1 in range(self.n_candidates):
            for cand2 in range(self.n_candidates):
                if cand1 != cand2:
                    for cand3 in range(self.n_candidates):
                        if cand3 not in (cand1, cand2):
                            curr_value = p_matrix[cand2][cand3]
                            new_value = min(
                                p_matrix[cand2][cand1],
                                p_matrix[cand1][cand3],
                            )
                            if new_value > curr_value:
                                p_matrix[cand2][cand3] = new_value
        return p_matrix

    def rank_p(self, p_matrix: np.ndarray):
        """
        Given a p matrix, return the ranking of the candidates.
        """
        ranking = np.zeros(self.n_candidates)
        for i in range(self.n_candidates):
            for j in range(self.n_candidates):
                if i != j:
                    if p_matrix[i][j] > p_matrix[j][i]:
                        ranking[i] += 1
        return ranking


if __name__ == "__main__":
    profile = np.array(
        [
            [0, 1, 2, 3],
            [0, 2, 3, 1],
            [2, 0, 1, 3],
            [3, 1, 2, 0],
            [3, 1, 2, 0],
        ]
    )
    profile = Profile(profile)
    d = profile.get_d_matrix()
    p = profile.get_p_matrix(d)
    print(p)
    ranking = profile.rank_p(p)
    print(ranking)

File: ese3090/beatpath/src/beatpath.rs
use itertools::Itertools;
use std::cmp;

fn index(n: usize, row: usize, col: usize) -> usize {
    row * n + col
}

fn get_d_matrix(profile: Vec<Vec<i32>>) -> Vec<i32> {
    let n_candidates = profile[0].len();
    let n_voters = profile.len();

    // Initialize 1D d_matrix with size n_candidates * n_candidates
    let mut d_matrix = vec![0; n_candidates * n_candidates];

    for comb in (0..n_candidates).combinations(2) {
        let a = comb[0];
        let b = comb[1];
        for i in 0..n_voters {
            if profile[i][a] < profile[i][b] {
                d_matrix[index(n_candidates, a, b)] += 1;
            } else {
                d_matrix[index(n_candidates, b, a)] += 1;
            }
        }
    }

    for i in 0..n_candidates {
        for j in 0..n_candidates {
            if i == j {
                d_matrix[index(n_candidates, i, j)] = 0;
            } else if d_matrix[index(n_candidates, i, j)] > d_matrix[index(n_candidates, j, i)] {
                d_matrix[index(n_candidates, j, i)] = 0;
            }
        }
    }

    d_matrix
}

fn get_p_matrix(d_matrix: Vec<i32>, n_candidates: usize) -> Vec<i32> {
    let mut p_matrix = vec![0; n_candidates * n_candidates];

    for cand1 in 0..n_candidates {
        for cand2 in 0..n_candidates {
            if cand1 != cand2 {
                let strength = d_matrix[index(n_candidates, cand1, cand2)];
                if strength > d_matrix[index(n_candidates, cand2, cand1)] {
                    p_matrix[index(n_candidates, cand1, cand2)] = strength;
                }
            }
        }
    }

    for cand1 in 0..n_candidates {
        for cand2 in 0..n_candidates {
            if cand1 != cand2 {
                for cand3 in 0..n_candidates {
                    if cand3 != cand1 && cand3 != cand2 {
                        let curr_value = p_matrix[index(n_candidates, cand2, cand3)];
                        let new_value = cmp::min(
                            p_matrix[index(n_candidates, cand1, cand3)],
                            p_matrix[index(n_candidates, cand2, cand1)],
                        );
                        if new_value > curr_value {
                            p_matrix[index(n_candidates, cand2, cand3)] = new_value;
                        }
                    }
                }
            }
        }
    }

    p_matrix
}

fn rank_p(p_matrix: Vec<i32>, n_candidates: usize) -> Vec<i32> {
    let mut ranking = vec![0; n_candidates];

    for i in 0..n_candidates {
        for j in 0..n_candidates {
            if i != j {
                if p_matrix[index(n_candidates, i, j)] < p_matrix[index(n_candidates, j, i)] {
                    ranking[i] += 1;
                }
            }
        }
    }

    ranking
}

pub fn compute_beatpath_ranking(profile: Vec<Vec<i32>>) -> Vec<i32> {
    let n_candidates = profile[0].len();
    let d_matrix = get_d_matrix(profile);
    let p_matrix = get_p_matrix(d_matrix, n_candidates);
    let ranking = rank_p(p_matrix, n_candidates);

    ranking
}
